src/Storages/MergeTree/MergeTreeDataPartWriterWide.cpp

```diff
@@ -607,6 +614,11 @@ void MergeTreeDataPartWriterWide::fillIndexGranularity(size_t index_granularity_
 
 void MergeTreeDataPartWriterWide::adjustLastMarkIfNeedAndFlushToDisk(size_t new_rows_in_last_mark)
 {
+    /// We don't want to split already written granules to smaller
+    if (rows_written_in_last_mark > new_rows_in_last_mark)
```

Comment:
Additional self-check. We have quite a lot of them, but I think it's better than possible data corruption.